pkg/rotation/reconciler.go

```diff
@@ -0,0 +1,346 @@
+		return nil, err
+	}
+	store, err := k8s.New(kubeClient, crdClient, nodeName, 60*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Reconciler{
+		store:                store,
+		ctrlReaderClient:     c,
+		ctrlWriterClient:     c,
+		scheme:               s,
+		providerVolumePath:   providerVolumePath,
+		rotationPollInterval: rotationPollInterval,
+		providerClients:      make(map[string]*secretsstore.CSIProviderClient),
+	}, nil
+}
+
+// Run starts the rotation reconciler
+func (r *Reconciler) Run(stopCh <-chan struct{}) {
+	log.Infof("starting rotation reconciler with poll interval: %s", r.rotationPollInterval)
+
+	ticker := time.NewTicker(r.rotationPollInterval)
+	defer ticker.Stop()
+
+	if err := r.store.Run(stopCh); err != nil {
+		log.Fatalf("failed to run informers for rotation reconciler, err: %+v", err)
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-ticker.C:
+			log.Infof("rotation reconcile started")
+			// The spc pod status informer is configured to do a filtered list watch of spc pod statuses
+			// labeled for the same node as the driver. LIST will only return the filtered results.
+			spcpsList, err := r.store.ListSecretProviderClassPodStatus()
+			if err != nil {
+				log.Errorf("[rotation] failed to list secret provider class pod status for node, err: %+v", err)
+				continue
+			}
+			for _, spcps := range spcpsList {
+				if err = r.reconcile(context.Background(), spcps); err != nil {
+					log.Errorf("[rotation] failed to reconcile spc %s/%s for pod %s/%s", spcps.Namespace,
+						spcps.Status.SecretProviderClassName, spcps.Namespace, spcps.Status.PodName)
+				}
+			}
+			log.Infof("rotation reconcile completed")
+		}
+	}
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, spcps *v1alpha1.SecretProviderClassPodStatus) error {
+	spcName, spcNamespace := spcps.Status.SecretProviderClassName, spcps.Namespace
+
+	// get the secret provider class the pod status is referencing from informer cache
+	spc, err := r.store.GetSecretProviderClass(spcName, spcNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get secret provider class %s/%s, err: %+v", spcNamespace, spcName, err)
+	}
+	paramsJSON, err := json.Marshal(spc.Spec.Parameters)
+	if err != nil {
+		return fmt.Errorf("failed to marshal parameters, err: %+v", err)
+	}
+	permissionJSON, err := json.Marshal(permission)
+	if err != nil {
+		return fmt.Errorf("failed to marshal permission, err: %+v", err)
+	}
+	// get pod from informer cache
+	podName, podNamespace := spcps.Status.PodName, spcps.Namespace
+	pod, err := r.store.GetPod(podName, podNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get pod %s/%s, err: %+v", podNamespace, podName, err)
+	}
+
+	// check if the volume pertaining to the current spc is using nodePublishSecretRef for
+	// accessing external secrets store
+	var nodePublishSecretRef *v1.LocalObjectReference
+	for _, vol := range pod.Spec.Volumes {
+		if vol.CSI.Driver != "secrets-store.csi.k8s.io" {
```

Comment:
do we need to check if the volume is csi here?

Reply:
`vol.CSI` checks only if the CSI volume is present and is for the `secrets-store.csi.k8s.io` driver.